File: webapp/src/lib/calibration.ts
/**
 * Calibration formulas that convert raw LabJack volts into engineering units.
 *
 * Calibrations live in the `calibrations` map of a LabJack config in the KV bucket
 * `avenabox`, keyed by channel number. The archiver stores the active one in each
 * Parquet file's metadata, and the exporter applies it to the CSV
 * `calibrated_value` column (see `rust-ljm/src/calibration.rs`). The dashboard uses
 * this module to clean up edited calibrations and to preview calibrated values.
 *
 * @module
 */

/**
 * Calibration formula plus an optional name, tagged by `type`.
 *
 * Same JSON shape as the Rust `CalibrationSpec`, for example
 * `{"id":"tp3505","type":"linear","a":2.0,"b":-1.0}`.
 */
export type CalibrationSpec =
  /** Leaves raw values unchanged. */
  | {
      /** Optional name; the exporter writes it in the `calibration_id` CSV column. */
      id?: string;
      /** Formula tag. */
      type: "identity";
    }
  /** Applies `a * raw + b`. */
  | {
      /** Optional name; the exporter writes it in the `calibration_id` CSV column. */
      id?: string;
      /** Formula tag. */
      type: "linear";
      /** Slope, in output units per volt. */
      a: number;
      /** Offset, in output units. */
      b: number;
    }
  /** Applies `coeffs[i] * raw ** i` and sums each term. */
  | {
      /** Optional name; the exporter writes it in the `calibration_id` CSV column. */
      id?: string;
      /** Formula tag. */
      type: "polynomial";
      /** Coefficients in ascending power order; `coeffs[0]` is the constant term. */
      coeffs: number[];
    };

/**
 * Converts a partial or unknown calibration object into a valid spec.
 *
 * Rules, by `type`:
 *
 * - Missing input or a non-string `type`: unnamed identity.
 * - `linear`: `a` and `b` are kept if they are finite numbers, otherwise `a`
 *   becomes 1 and `b` becomes 0. Numeric strings are not accepted.
 * - `polynomial`: each coefficient is converted with `Number()` and non-finite
 *   results are dropped. An empty result becomes `[0, 1]`, the identity polynomial.
 * - Any other `type`: identity.
 *
 * `id` is carried over in every case except the first. Returns a new object and does
 * not modify `raw`.
 *
 * @param raw - Calibration as read from config or form state, possibly partial.
 * @returns A spec that {@link applyCalibration} can evaluate.
 *
 * @example
 * ```ts
 * normalizeCalibration({ type: "linear", a: 2 });
 * // { id: undefined, type: "linear", a: 2, b: 0 }
 * normalizeCalibration({ type: "polynomial", coeffs: [] });
 * // { id: undefined, type: "polynomial", coeffs: [0, 1] }
 * ```
 */
export function normalizeCalibration(
  raw?: Partial<CalibrationSpec> | null
): CalibrationSpec {
  if (!raw || typeof raw.type !== "string") {
    return { type: "identity" };
  }

  if (raw.type === "linear") {
    return {
      id: raw.id,
      type: "linear",
      a: Number.isFinite(raw.a as number) ? Number(raw.a) : 1,
      b: Number.isFinite(raw.b as number) ? Number(raw.b) : 0,
    };
  }

  if (raw.type === "polynomial") {
    const coeffs = Array.isArray(raw.coeffs)
      ? raw.coeffs.map((value) => Number(value)).filter((value) => Number.isFinite(value))
      : [];
    return {
      id: raw.id,
      type: "polynomial",
      coeffs: coeffs.length > 0 ? coeffs : [0, 1],
    };
  }

  return { id: raw.id, type: "identity" };
}

/**
 * Applies a calibration formula to one raw sample value.
 *
 * A non-finite `raw` (`NaN` for a skipped sample, or an infinity) is returned
 * unchanged without evaluating the formula. The spec is used as given; pass it
 * through {@link normalizeCalibration} first if it may be incomplete.
 *
 * @param spec - Calibration to apply.
 * @param raw - Raw reading, in volts.
 * @returns The calibrated value. An empty polynomial evaluates to 0.
 *
 * @example
 * ```ts
 * applyCalibration({ type: "linear", a: 2, b: -1 }, 3);          // 5
 * applyCalibration({ type: "polynomial", coeffs: [1, 0, 2] }, 3); // 1 + 2 * 9 = 19
 * ```
 */
export function applyCalibration(spec: CalibrationSpec, raw: number): number {
  if (!Number.isFinite(raw)) {
    return raw;
  }

  if (spec.type === "linear") {
    return spec.a * raw + spec.b;
  }

  if (spec.type === "polynomial") {
    return spec.coeffs.reduce((acc, coeff, idx) => acc + coeff * raw ** idx, 0);
  }

  return raw;
}

File: rust-ljm/src/calibration.rs
//! Calibration formulas stored with archived channel data.
//!
//! Calibrations come from the `calibrations` map in the dashboard sensor settings.
//! The archiver writes the channel's [`CalibrationSpec`] as JSON into each Parquet
//! file's key-value metadata under the key `calibration`. The exporter reads that
//! metadata back and applies the formula while generating CSV rows, so an export
//! uses the calibration that was in force when the file was written.
//!
//! The JSON shape is the formula's fields plus an optional `id`, tagged by `type`:
//!
//! ```text
//! {"type":"identity"}
//! {"id":"tp3505","type":"linear","a":2.0,"b":-1.0}
//! {"type":"polynomial","coeffs":[0.1,2.0,0.03]}
//! ```

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case")]
/// Formula used to convert a raw LabJack reading into a calibrated value.
///
/// Serialized with an internal `type` tag in snake_case (`identity`, `linear`,
/// `polynomial`).
pub enum CalibrationFormula {
    /// Leaves the raw value unchanged.
    Identity,
    /// Applies `a * raw + b`, where `a` is the slope and `b` the offset.
    Linear { a: f64, b: f64 },
    /// Applies the polynomial `sum(coeffs[i] * raw.powi(i))`.
    ///
    /// `coeffs` is in ascending power order, so `coeffs[0]` is the constant term.
    /// An empty list evaluates to `0.0`.
    Polynomial { coeffs: Vec<f64> },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
/// Named calibration definition for one channel.
///
/// Deserialized from the dashboard sensor settings by the archiver and from Parquet
/// metadata by the exporter. The default is an unnamed identity calibration.
pub struct CalibrationSpec {
    /// Optional identifier. The exporter writes it in the `calibration_id` CSV
    /// column, or `identity` when it is `None` (see [`Self::id_or_default`]).
    pub id: Option<String>,
    /// Formula; its `type` tag and fields sit at the top level of the JSON object
    /// next to `id` (serde `flatten`).
    #[serde(flatten)]
    pub formula: CalibrationFormula,
}

// The archiver compiles this module too but only stores specs, so it never calls
// these methods.
#[allow(dead_code)]
impl CalibrationSpec {
    /// Applies the calibration formula to one raw sample value.
    ///
    /// NaN and infinite inputs propagate through the arithmetic unchanged.
    ///
    /// # Arguments
    ///
    /// * `raw` - Raw sample value as stored in the archive.
    ///
    /// # Returns
    ///
    /// The calibrated value.
    ///
    /// # Examples
    ///
    /// ```text
    /// Linear { a: 2.0, b: -1.0 }             applied to 3.0 == 5.0
    /// Polynomial { coeffs: [1.0, 0.0, 2.0] } applied to 3.0 == 1 + 2 * 9 == 19.0
    /// ```
    pub fn apply(&self, raw: f64) -> f64 {
        match &self.formula {
            CalibrationFormula::Identity => raw,
            CalibrationFormula::Linear { a, b } => a * raw + b,
            CalibrationFormula::Polynomial { coeffs } => coeffs
                .iter()
                .enumerate()
                .map(|(idx, coeff)| coeff * raw.powi(idx as i32))
                .sum(),
        }
    }

    /// Returns the configured calibration id, or `identity` when unnamed.
    ///
    /// The fallback is the string `identity` even if the formula is not the
    /// identity.
    pub fn id_or_default(&self) -> &str {
        self.id.as_deref().unwrap_or("identity")
    }
}

impl Default for CalibrationSpec {
    /// Creates an unnamed identity calibration.
    fn default() -> Self {
        Self {
            id: None,
            formula: CalibrationFormula::Identity,
        }
    }
}
